### PR TITLE
Expose mechanisms in the high-level API

### PR DESCRIPTION
**Pull request by cipherboy**
This creates a new class, Mechanism, for inquiring
information about a mechanism. This includes support
for RFCs 5587 and 5801. As Mechanism derives from OID,
it is compatible with all places that accept a mech
by OID.

Points of discussion (mechs):
  - ~Fallback in `mech_name()`: what mechs should we support? We could hard-code the three MIT/KRB5 OIDs (krb5/iakerb/spnego) with respective names for the fallback when RFC5801 is missing, but I'd like thoughts on that.~
  - ~Am I missing functions that significantly depend on a OID?~
  - Do we want other high-level calls to default to exposing Mechanisms vs. raw OIDs? e.g., `init_sec_context`?
  - ~Revising __init__() -- I wasn't able to get a call to `super().__init__` to work like was done with creds.   Thoughts would be appreciated.~

Regarding mech_attrs, I'm not yet sure how I want to handle that. `display_mech_attr` is about the only useful function I can see that depends on mech_attrs, and querying for a list of mech_attrs isn't trivial--It'd likely involve a call to `indicate_mechs`+`inquire_attrs_for_mechs`. I'm not really seeing a use for having mech_attrs as a class. Thoughts?

### CODE REVIEW

File: gssapi/mechs.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        """
+        if rfc5801 is None:
+            return str(roids.OID(self))
+        self._query_saslname()
+        return self._sasl_name.mech_name.decode('UTF-8')
+
+    @property
+    def sasl_name(self):
+        """
+        Get the SASL name for the mechanism; depends on RFC 5801
```
Comment (cipherboy) [inline gssapi/mechs.py]: (done)

**Comment by cipherboy** [comment]
Thanks for the detailed review and the help tracking those issues down yesterday. :)

I added `inquire_mechs_for_name()` support as a missing mechanism-related function.

TODO:
  - Python 2/3 `str` handling.
  - Fallback mech names for pretty printing
  - `mech_attrs`

**Comment by frozencemetery** [comment]
Hardcoding anything with a standard is fine in principle, but kind of defeats the point of calling into the GSSAPI there at all... I'm not decided on that.

**Comment by cipherboy** [comment]
On second thought, I don't know if there is value to this. Both MIT and Heimdal support the RFC 5801 extensions, so mech_name will be available under both.

File: gssapi/mechs.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        return self._query_saslname.mech_name.decode('UTF-8')
+
+    @property
+    def description(self):
+        """
+        Get the description of the mechanism; depends on RFC 5801
+
+        :requires-ext:`rfc5801`
+        """
+        self._query_saslname()
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: redundant call

File: gssapi/mechs.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        return self._query_attrs.mech_attrs
+
+    @classmethod
+    def all_mechs(cls):
+        """
+        all_mechs()
+        Get a list of all mechanisms supported by GSSAPI
+        """
+        mechs = rmisc.indicate_mechs()
+        return [cls(mech) for mech in mechs]
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: collapse to `(cls(mech) for mech in metchs)`

File: gssapi/mechs.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        Raises:
+            GSSError
+
+        :requires-ext:`rfc5801`
+        """
+        if rfc5801 is None:
+            raise NotImplementedError("Your GSSAPI implementation does not "
+                                      "have support for RFC 5801")
+        n = name
+        if type(n) is not bytes:
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: use `isinstance`

File: gssapi/mechs.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+    def __bytes__(self):
+        base = self.dotted_form
+        if rfc5801 is not None:
+            base = self._saslname.mech_name
+
+        if issubclass(str, six.text_type):
+            base = bytes(base, _utils._get_encoding())
+        else:
+            base = bytes(base)
```
Comment (cipherboy) [inline gssapi/mechs.py]: Since `dotted_form` returns a string, this needs to be cast to `bytes` for python2 compatibility. However, python3 forces two arguments to `bytes`, one of which is the encoding. Suggestions on how to make this better?

File: gssapi/raw/oids.pyx [hunk 6]
```diff
@@ -120,11 +132,15 @@ cdef class OID:
         if self._free_on_dealloc:
             free(self.raw_oid.elements)
 
-    def __bytes__(self):
+    @property
+    def byte_form(self):
```
Comment (DirectXMan12) [inline gssapi/raw/oids.pyx]: err... why is this necessary?  Why not just leave it as `__bytes__`?

File: gssapi/mechs.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        if issubclass(str, six.text_type):
+            # Python 3 -- we should return unicode
+            return self.__bytes__().decode(_utils._get_encoding())
+        else:
+            return self.__bytes__()
+
+    def __unicode__(self):
+        return self.__bytes__().decode(_utils._get_encoding())
+
+    def __bytes__(self):
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: hmm... I'd just leave `__bytes__` as-is.  Let in inherit from the low-level API for compatibility with OIDs -- some things might expect OID bytes when calling `__bytes__` -- we shouldn't change that part of the API just because you're working Mechanisms.

Comment (DirectXMan12) [inline gssapi/mechs.py]: then put the "string form as bytes" in a separate private method (like `_byte_desc`).

**Comment by cipherboy** [comment]
Okie dokie, this is updated as well. In retrospect, not changing the `__bytes__` interface was probably a good idea. :)

File: gssapi/mechs.py (hunk 5 above)
Comment (DirectXMan12) [reply gssapi/mechs.py]: it should be fine as

```python
if isinstance(base, six.text_type):
    base = base.encode(_utils._get_encoding())
```

File: gssapi/mechs.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DirectXMan12) [reply gssapi/mechs.py]: while you're doing updates, remove the "; depends on RFC 5801" -- just the `requires-ext` tag below is fine.

File: gssapi/mechs.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+            GSSError
+
+        :requires-ext:`rfc5801`
+        """
+        if rfc5801 is None:
+            raise NotImplementedError("Your GSSAPI implementation does not "
+                                      "have support for RFC 5801")
+        n = name
+        if not isinstance(n, bytes):
+            n = str(n).encode()
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: The pattern we generally use here is

```python
if isinstance(name, six.text_type):
    name = name.encode(_utils._get_encoding())
```

See the `gssapi/names.py` (at the bottom) for example

File: gssapi/tests/test_high_level.py [hunk 9]
```diff
@@ -369,6 +370,57 @@ def test_add_with_impersonate(self):
...
+            s.shouldnt_be_empty()
+
+    @ktu.gssapi_extension_test('rfc5801', 'RFC 5801: SASL Names')
+    def test_sasl_properties(self):
+        mechs = gssmechs.Mechanism.all_mechs()
+        for mech in mechs:
+            s = str(mech)
+            s.shouldnt_be_empty()
+            s.should_be_a(str)
+            s[0].shouldnt_be('<')
```
Comment (DirectXMan12) [inline gssapi/tests/test_high_level.py]: this is a fairly fragile line here.  I'd just remove it, unless the SASL names are *guaranteed* to not begin with "<"...

File: gssapi/tests/test_high_level.py [hunk 10]
```diff
@@ -369,6 +370,57 @@ def test_add_with_impersonate(self):
...
+        for mech in mechs:
+            attrs = mech.attrs
+            known_attrs = mech.known_attrs
+
+            for attr in attrs:
+                i = list(gssmechs.Mechanism.from_attrs(m_desired=[attr]))
+                e = list(gssmechs.Mechanism.from_attrs(m_except=[attr]))
+
+                count = len(i) + len(e)
+                count.should_be(c)
```
Comment (DirectXMan12) [inline gssapi/tests/test_high_level.py]: `(len(from_desired) + len(from_except)).should_be(c)` so we get a good error message like `len(from_desired) + len(from_except) should be 37`.

File: gssapi/raw/oids.pyx [hunk 11]
```diff
@@ -61,6 +61,9 @@ cdef class OID:
         self._free_on_dealloc = True
         return 0
 
+    def _copy_oid(self, OID other):
```
Comment (DirectXMan12) [inline gssapi/raw/oids.pyx]: this doesn't seem to be needed any more

File: gssapi/mechs.py [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    mechanisms and can be used with any calls.
+
+    It inherits from the low-level GSSAPI :class:`~gssapi.raw.oids.OID` class,
+    and thus can be used with both low-level and high-level API calls.
+    """
+    def __new__(cls, cpy=None, elements=None):
+        return super(Mechanism, cls).__new__(cls, cpy, elements)
+
+    @property
+    def names(self):
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: this should probably be `name_types`, to be clearer -- we use the term `name_type` elsewhere in the high-level API

File: gssapi/mechs.py [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+        Get the SASL name for the mechanism
+
+        :requires-ext:`rfc5801`
+        """
+        return self._saslname.sasl_mech_name.decode('UTF-8')
+
+    @property
+    def description(self):
+        """
+        Get the description of the mechanism; depends on RFC 5801
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: you can remove `depends on RFC 5801`, since you have the `requires-ext` tag.

File: gssapi/mechs.py [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+
+        :requires-ext:`rfc5801`
+        """
+        return self._saslname.mech_description.decode('UTF-8')
+
+    @property
+    def known_attrs(self):
+        """
+        Get the known attributes of the mechanism; depends on RFC 5587
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: You should document what form an "attribute" takes.

File: gssapi/mechs.py (hunk 13 above)
Comment (DirectXMan12) [reply gssapi/mechs.py]: ditto for the cases below

File: gssapi/mechs.py [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+        Get the attributes of the mechanism; depends on RFC 5587
+
+        :requires-ext:`rfc5587`
+        """
+        return self._attrs.mech_attrs
+
+    @classmethod
+    def all_mechs(cls):
+        """
+        all_mechs()
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: no need for the explicit signature here.

Comment (DirectXMan12) [reply gssapi/mechs.py]: ditto below

File: gssapi/mechs.py [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+        m = rfc5801.inquire_mech_for_saslname(name)
+
+        return cls(m)
+
+    @classmethod
+    def from_attrs(cls, m_desired=None, m_except=None, m_critical=None):
+        """
+        from_attrs
+        Get a generator of mechanisms supporting the specified attributes. See
+        RFC 5587's indicate_mechs_by_attrs for more information.
```
Comment (DirectXMan12) [inline gssapi/mechs.py]: link to `indicate_mechs_by_attrs` using the correct documentation tag (IIRC, it's something like `:func:`).

File: gssapi/mechs.py (hunk 13 above)
Comment (cipherboy) [reply gssapi/mechs.py]: Sorry, should've caught those sooner if I were thinking when I updated the PR.

File: gssapi/mechs.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+        ([OID])
+
+        :requires-ext:`rfc5587`
+        """
+        return self._attrs.known_mech_attrs
+
+    @property
+    def attrs(self):
+        """
+        Get the attributes of the mechanism; returns a set of OIDs ([OID])
```
Comment (cipherboy) [inline gssapi/mechs.py]: Leaving this open as I'm not sure if this is the desired format.

Comment (cipherboy) [reply gssapi/mechs.py]: Per IRC discussion with @DirectXMan12 this appears to be in the correct format.

**Comment by cipherboy** [comment]
@DirectXMan12 -- any new feedback would be appreciated. :)

**Comment by DirectXMan12** [comment]
/me attempts to page this back into RAM

**Comment by DirectXMan12** [comment]
Looks like my last comments were just around docs, so assuming past me's review was fairly thorough (I recall it being so), I'm going to merge this.
